### ISSUE TITLE
[BUG]  Deadlock in OpenMP Parallel Loop When Invoking LAPACK Functions in Cython>=3.0

### ISSUE THREAD

**Issue by ameli** [cython/cython#6107]
### Describe the bug

This issue specifically pertains to Cython versions 3.0 and above. It involves a deadlock encountered when calling LAPACK functions, such as [`dstev`](https://docs.huihoo.com/doxygen/lapack/3.7.0/dc/dd2/group__double_o_t_h_e_reigen_gaaa6df51cfd92c4ab08d41a54bf05c3ab.html), within an OpenMP parallel loop in C++.

**Accessing LAPACK in C++:**

The `scipy.linalg.cython_lapack` module offers a Cython interface to several LAPACK functions, which can be `cimport`ed. These Cython declarations can be directly called from C++ through [C++ public declarations](https://docs.cython.org/en/latest/src/userguide/external_C_code.html#c-public-declarations).

**Issue Detail:**

The setup involves C++ code (in `c_module.h` and `c_module.cpp` and interface `c_module.pxd`), which is wrapped in Cython in `module.pyx`. The LAPACK function `dstev` is made accessible from Scipy's Cython interface to C++ via `lapack_api.pxi`. This configuration works as expected unless Cython version 3.0 or later is used.

The issue arises in a parallel loop (within `c_module.cpp` in the `compute()` function), where the loop fails to exit upon completion. Specifically, after all iterations are performed and `dstev` is correctly called in each iteration, the program halts indefinitely without exiting the loop, without any warning or CPU usage.

**Conditions for Issue Reproduction:**

The issue only manifests under all of the following conditions:

1. OpenMP is set to use more than one thread, **AND**
2. Cython is version 3.0 or above, **AND**
3. Scipy is version 1.10.1 or below, **AND**
4. A LAPACK function, such as `dstev`, is called.

The issue does not occur if any of the above conditions are not met. Specifically, it cannot be reproduced if:

1. OpenMP uses only one thread, **OR**
2. Cython is below version 3.0, **OR**
3. Scipy is above version 1.10.1, **OR**
4. LAPACK functions are not called at all.

**Suspected Cause:**

Scipy employs OpenBLAS for its LAPACK implementation, which I suspect may request multiple threads, thereby interfering with parent OpenMP loops. Regrettably, there is no scipy's interface to OpenBLAS to limit the number of threads OpenBLAS requests. Though this seems an [OpenBLAS issue](https://github.com/OpenMathLib/OpenBLAS/wiki/Faq#how-can-i-use-openblas-in-multi-threaded-applications), it peculiarly only arises with Cython versions 3.0 and above. All compilations with Cython below 3.0 function correctly under any scenario, including various runtime thread counts and LAPACK function calls.

Despite the narrow conditions under which this issue occurs, it is critically limiting, affecting the use of Cython 3.0 or newer with specific LAPACK implementations.

I look forward to any insights or resolutions you may provide regarding this matter.

### Code to reproduce the behaviour:

```python
# file: setup.py

from setuptools import setup
from Cython.Build import cythonize
from distutils.extension import Extension

extensions = [
    Extension("module",
              ["module.pyx", "c_module.cpp"],
              language='c++',
              extra_compile_args=['-O3', '-fopenmp'],
              extra_link_args=['-fopenmp'],
              define_macros=[("CYTHON_EXTERN_C", 'extern "C"')])
]

setup(
    name="trace",
    ext_modules=cythonize(extensions),
)
```

```cpp
// file: c_module.h

#ifndef _module_h
#define _module_h

CYTHON_EXTERN_C
{
    void lapack_dstev(char* jobz, int* n, double* d, double* e, double* z,
                      int* ldz, double* work, int* info);
}

void compute(double* d, double* e, double* v, int n);

#endif
```

```cpp
// file: c_module.cpp

#include "./c_module.h"
#include "omp.h"

void _tridiag_eig(double* d, double* e, double* v, int n)
{
    char jobz = 'V';
    double *work = new double[2*n-2];

    int ldz = n;
    int info;

    // By commenting this line blow, the issue will not occur.
    lapack_dstev(&jobz, &n, d, e, v, &ldz, work, &info);

    delete[] work;
}

void compute(double* d, double* e, double* v, int n)
{
    int repeat = 10;

    // By setting number of threads to 1, the issue will not occur.
    // omp_set_num_threads(1);

    #pragma omp parallel for
    for (int i=0; i < repeat; ++i)
    {
        _tridiag_eig(d, e, v, n);
    }

    // Code never reaches this point at runtime.
}
```

```cython
# file: c_module.pxd

cdef extern from "c_module.h":
    void compute(double* d, double* e, double* v, int n) nogil
```

```cython
# file: lapack_api.pxi

from scipy.linalg.cython_lapack cimport dstev

cdef public void lapack_dstev(
        char* jobz, int* n, double* d, double* e, double* z, int* ldz,
        double* work, int* info) noexcept nogil:

    dstev(jobz, n, d, e, z, ldz, work, info)
```

```cython
# file: module.pyx

include "lapack_api.pxi"
from c_module cimport compute

cpdef void compute_wrapper(
        double[:] d, double[:] e, double[:] v, int n) noexcept nogil:

    cdef double* dp = &d[0];
    cdef double* ep = &e[0];
    cdef double* vp = &v[0];

    compute(dp, ep, vp, n);
```

```python
# file: main.py

import numpy as numpy
import module

n = 5
d = numpy.random.randn(n)
e = numpy.random.randn(n-1)
v = numpy.zeros((n*n, ), dtype=numpy.float64)

module.compute_wrapper(d, e, v, n)
print(d)
```

### Expected behaviour

OpenMP's for loop should exit the parallel region after the last iteration.

### OS

Ubuntu 22.04

### Python version

3.10.8

### Cython version

3.0.9

### Additional context

Note that to reproduce the issue, you need `scipy<=1.10`, which does not have wheel for python 3.12. So, please use python<3.12.

Also, the output of `python -c "import scipy; print(scipy.show_config())"` is

```
blas_mkl_info:
  NOT AVAILABLE
blis_info:
  NOT AVAILABLE
openblas_info:
    libraries = ['openblas', 'openblas']
    library_dirs = ['/usr/local/lib']
    language = c
    define_macros = [('HAVE_CBLAS', None)]
    runtime_library_dirs = ['/usr/local/lib']
blas_opt_info:
    libraries = ['openblas', 'openblas']
    library_dirs = ['/usr/local/lib']
    language = c
    define_macros = [('HAVE_CBLAS', None)]
    runtime_library_dirs = ['/usr/local/lib']
lapack_mkl_info:
  NOT AVAILABLE
openblas_lapack_info:
    libraries = ['openblas', 'openblas']
    library_dirs = ['/usr/local/lib']
    language = c
    define_macros = [('HAVE_CBLAS', None)]
    runtime_library_dirs = ['/usr/local/lib']
lapack_opt_info:
    libraries = ['openblas', 'openblas']
    library_dirs = ['/usr/local/lib']
    language = c
    define_macros = [('HAVE_CBLAS', None)]
    runtime_library_dirs = ['/usr/local/lib']
Supported SIMD extensions in this NumPy install:
    baseline = SSE,SSE2,SSE3
    found = SSSE3,SSE41,POPCNT,SSE42,AVX,F16C,FMA3,AVX2
    not found = AVX512F,AVX512CD,AVX512_KNL,AVX512_KNM,AVX512_SKX,AVX512_CLX,AVX512_CNL,AVX512_ICL
None

```

**Comment by da-woods** [comment]
I've install Scipy 1.10.1 in a virtual environment and I'm struggling to find "lapack_api.pxi" so I think there's some details I'm missing.

However - here's my working theory:

in `compute_wrapper` the `nogil` says that it *can* run without the GIL, not that it actually releases the GIL. You don't release the GIL and thus the whole thing runs with the GIL locked..

I wouldn't be surprised if it worked if you changed

```
cpdef void compute_wrapper(
        double[:] d, double[:] e, double[:] v, int n) noexcept:

    cdef double* dp = &d[0];
    cdef double* ep = &e[0];
    cdef double* vp = &v[0];

    with nogil:
        compute(dp, ep, vp, n);
```

to explicitly release the GIL.

I'm not sure what's going on beyond that (since I think the mystery .pxi file is important). But it might be that in Cython 3 the implementation of `lapack_dstev` has added an exception check somewhere which tries to reacquire the GIL.

**Comment by ameli** [comment]
I added `lapack_api.pxi` to the original post. Thanks for noting the missing file.

**Comment by mattip** [comment]
This is happening on linux, so probably not related directly to https://github.com/scipy/scipy/issues/20294 which is windows-specific

**Comment by da-woods** [comment]
So:

In that version of scipy `dstev` is written as:

```
cdef void dstev(char *jobz, int *n, d *d, d *e, d *z, int *ldz, d *work, int *info) nogil
```

(I'm fairly sure it has `noexcept` in more recent versions of Scipy). Cython 3 interprets the definition without `noexcept` to mean "might throw a Python exception". Therefore, the call to `dstev` generates to code

```
__pyx_f_5scipy_6linalg_13cython_lapack_dstev(__pyx_v_jobz, __pyx_v_n, __pyx_v_d, __pyx_v_e, __pyx_v_z, __pyx_v_ldz, __pyx_v_work, __pyx_v_info); if (unlikely(__Pyx_ErrOccurredWithGIL()))
```

i.e. it *always* re-acquires the GIL in order to check for an exception.

As I said earlier, `compute_wrapper` doesn't release the GIL - it can simply run without the GIL. Therefore the deadlock is that all your OpenMP threads are trying to acquire the GIL when something else already holds it, hence the block.

If you modify `compute_wrapper` as I suggested to explicitly release the GIL then it works correctly.

We ran into a few issues like this with the release of Cython 3. The one differs because the parallel loop isn't Cython-generated, so it can't be fixed by changing the GIL handling from Cython.
